src/base/bittorrent/torrentcreator.cpp

```diff
@@ -204,14 +205,22 @@ void TorrentCreator::run()
             entry["info"]["source"] = m_params.source.toStdString();
 
         checkInterruptionRequested();
+        if (!m_params.savePath.isEmpty())
```

Comment:
Is it intended to allow empty save path?

Reply:
Yes, whoever creates a torrent via the API doesn't necessarily want the torrent on the disk. This allows the workflow "create a torrent" -> "retrieve the torrent via api" -> "push the torrent via api some other servers" -> "let the transfer happen between all servers". In this use-case, temporary .torrent files are just an additional un-needed hassle to manage. 

The current PR came from my desire to implement Bittorrent as a wire synchronization protocol between storage elements in [rucio](https://github.com/rucio/rucio)

Reply:
In reality, the workflow will be more complicated. The workflow which I described will only used for files for which we don't (yet) know the bittorrent v2 merkle root and piece layers. But this avoids having to implement a custom agent to compute this data and communicate it to Rucio if the bittorrent client already runs on the storage server anyway. I proposed a related change to deluge too (https://github.com/deluge-torrent/deluge/pull/430)

Reply:
So I believe it isn't needed to create torrent file if client need the data?
I would omit additional parameter and use only savePath to determine whether file or data to produce.

Reply:
In this case, if the client will try to fetch the file via the `getFile` action, it will open the file on disk? And fail if the file was removed ?

Reply:
> In this case, if the client will try to fetch the file via the `getFile` action, it will open the file on disk? And fail if the file was removed ?

Sorry, I don't understand what you mean.

I meant, wouldn't it look more logical, producing one thing, either data or a file? Then only one parameter (savePath) would be enough to control this aspect. I.e., if the path is specified, then save the data to a file, and if the path is not specified (empty), then return the data.

Reply:
(I talk about TorrentCreator class and by "client", I mean the external code that interacts with it.)

Reply:
I meant this Action : 
https://github.com/qbittorrent/qBittorrent/pull/19498/files#diff-def5c88a17643578a225e609c1350b7721847ae85022ed0b3d4ac38a0bf96980R258

It allows to fetch the content of the .torrent file. It's possible to make it work without the new parameter by opening the file whenever the api call is done, but it's less robust, because the file may get deleted in the meantime.

Reply:
> It's possible to make it work without the new parameter

Above I talked about the fact that TorrentCreator should either return data within result or create torrent file depending on savePath provided. If you don't want to create torrent file on disk you can just pass empty savePath to TorrentCreator constructor so no additional parameter is really needed.
How does what I'm suggesting contradict what you want to do in the API controller?

Reply:
As for the API controller itself.
Is it such a good idea to store all this data in memory, given that they have an indefinite lifetime? Perhaps I am mistaken, since I still haven't reviewed it in detail. What is the life cycle of the "torrent creation" job? Are there any time limits, or a limit on the number of simultaneously existing jobs? Or are they deleted only at the request of the Web API client and at the end of the session?

Reply:
I didn't meant it as a contradiction. I definitely wasn't clear enough, but I was asking for your opinion on how to handle the api endpoint which fetches the content of the created torrent in the case that the content is not returned and stored in memory. I propose to implement the change that you suggested + make `getFileAction` open the file and serve its content if the file still exist on disk; but return a failure if the file was removed. Is it a good idea to do it like that. 

As for torrent creation jobs. You quite correct. The jobs are bound to a session and they should get destroyed when the session is closed, but by default (I believe so?) bittorent keeps the session indefinitely. Do you have a suggested solution? I was thinking initially to keep track of jobs in the base/bittorrent/session singleton. But I'm not sure it's a good idea. Any suggestions?

Reply:
> Any suggestions?

Did you consider the search controller as I suggested before?

Reply:
> The jobs are bound to an api/session and they should get destroyed when the api/session is closed, but by default (I believe so?) bittorent keeps the session indefinitely.

I talked about Web Session. It is created during the Web user's login and lives until the user logs out or until it expires (if the user has been inactive for some time). API controller instances live inside Web Session.

Reply:
>I propose to implement the change that you suggested + make getFileAction open the file and serve its content if the file still exist on disk; but return a failure if the file was removed.

One excludes the other.
My proposed changes (regarding the TorrentCreator class) useful if you still want to obtain the data and store it in memory. If the API controller does not need to store data in memory, then simply revert your current changes to TorrentCreator class.

Reply:
I added a separate TorrentCreationManager singleton. Also, i changed the verbs(action names) used on the api, but I chose to use create/status/result(singular)/delete compared to the ones you mentioned previously. It feels somehow strange to have a "start" and "stop" action for torrent creation. And you can only fetch one file result at a time. What do you think?

Reply:
> I chose to use create/status/result(singular)/delete

👍 


> It feels somehow strange to have a "start" and "stop" action for torrent creation.

I have never insisted on exact copying. The main thing is to catch the essence.


> I added a separate TorrentCreationManager singleton.

👎 
Well, I don't even want to look there. I believe in advance that this is unnecessary. (I don't want to sound arrogant, but my experience says so.)
Why do you need yet another manager? Isn't the controller itself enough for you to have "managing" logic there?

Reply:
:sob: this was a compete miss-understanding then, I thought that's what you were telling me to do in answer to 

> As for torrent creation jobs. You are correct. The jobs are bound to an api/session and they should get destroyed when the api/session is closed, but by default (I believe so?) bittorent keeps the session indefinitely. Do you have a suggested solution? I was thinking initially to keep track of jobs in the base/bittorrent/session singleton. But I'm not sure it's a good idea. Any suggestions?

When you said to check the SearchController. By storing the torrent creation tasks in a singleton rather than inside the controller, they are not bound to a web session anymore. Having them managed by a web session doesn't allow to enforce any limits on the number of total torrent creation tasks allowed globally in parallel: the user can always bypass the limit by logging in parallel to a new Web Session and submitting new tasks under the new session. 

That being said: I don't have any strong preferences for any of the two versions:
- I can rollback to the previous version, where everything was handled inside the controller, but had the limitations of impossibility to enforce limits + the limitation that creation tasks where not visible across 2 different WebSession (if we ever add the functionality to the web ui, the users will be surprised if the torrent creation jobs disappear after a disconnect).
- I can keep the new version with the Manager which has some advantages, but introduces a new singleton. This seems to be non-desired according to your last comment.

Reply:
@rcarpa 
As I said before I still didn't deal with it carefully so I can miss some important details.
Could you first provide some general description of how you think it should be designed?

Reply:
@glassez , for my own requirements, I just need a way to create a torrent file via the api. I really don't have any preferences concerning technical aspects of the implementation. I try to do my best to implement things in a way similar to what already exists and to follow your pointers to achieve a better-quality result, to fulfill the needs of the broader community of this project. However, it's my first contribution to the project and I don't have near enough high-level perspective to design it in a way which fits the long term vision of the project. It will be very useful if somebody with a long experience in the project will give it a thought and will tell what's desired and what's not desired.

Reply:
@rcarpa
Hmm... it seems that there are some aspects of this case that I didn't realize at first. I need to think about it. Then I will try to summarize some of my judgments so that we can continue the discussion more productively.